mllib/src/main/scala/org/apache/spark/ml/feature/TargetEncoder.scala

```diff
@@ -0,0 +1,447 @@
+ * probability of the target given that category (bin counting).
+ * When 'targetType' is configured to 'continuous', categories are encoded as the average
+ * of the target given that category (mean encoding)
+ *
+ * Parameter 'smoothing' controls how in-category stats and overall stats are weighted.
+ *
+ * @note When encoding multi-column by using `inputCols` and `outputCols` params, input/output cols
+ * come in pairs, specified by the order in the arrays, and each pair is treated independently.
+ *
+ * @see `StringIndexer` for converting categorical values into category indices
+ */
+@Since("4.0.0")
+class TargetEncoder @Since("4.0.0") (@Since("4.0.0") override val uid: String)
+  extends Estimator[TargetEncoderModel] with TargetEncoderBase with DefaultParamsWritable {
+
+  @Since("4.0.0")
+  def this() = this(Identifiable.randomUID("TargetEncoder"))
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setLabelCol(value: String): this.type = set(labelCol, value)
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setInputCols(values: Array[String]): this.type = set(inputCols, values)
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setOutputCols(values: Array[String]): this.type = set(outputCols, values)
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setHandleInvalid(value: String): this.type = set(handleInvalid, value)
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setTargetType(value: String): this.type = set(targetType, value)
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setSmoothing(value: Double): this.type = set(smoothing, value)
+
+  @Since("4.0.0")
+  override def transformSchema(schema: StructType): StructType = {
+    validateSchema(schema, fitting = true)
+  }
+
+  @Since("4.0.0")
+  override def fit(dataset: Dataset[_]): TargetEncoderModel = {
+    validateSchema(dataset.schema, fitting = true)
+
+    val feature_types = inputFeatures.map{
+      feature => dataset.schema(feature).dataType
+    }
+    val label_type = dataset.schema($(labelCol)).dataType
+
+    val stats = dataset
+      .select((inputFeatures :+ $(labelCol)).map(col).toIndexedSeq: _*)
+      .rdd.treeAggregate(
+        Array.fill(inputFeatures.length) {
+          Map.empty[Option[Double], (Double, Double)]
+        })(
+        (agg, row: Row) => if (!row.isNullAt(inputFeatures.length)) {
+          val label = label_type match {
+            case ByteType => row.getByte(inputFeatures.length).toDouble
+            case ShortType => row.getShort(inputFeatures.length).toDouble
+            case IntegerType => row.getInt(inputFeatures.length).toDouble
+            case LongType => row.getLong(inputFeatures.length).toDouble
+            case DoubleType => row.getDouble(inputFeatures.length)
```

Comment:
I would suggest make the casting happen before the aggregation (`dataset.select` in the above ) to simplify the process

Reply:
done